Enhance/precision num

edlins opened the pull request:
Fixes #198 .

Changes proposed in this pull request:
- PrecisionNum introduction
- add Num numOf(String, int)
- more tests in NumTest
- add TestUtils.assertIndicatorEquals() with delta for inexact matching
- logging
- rename DoubleNumVsBigDecimalNum to CompareNumTypes, add error percentage reporting

- [ ] added an entry to the unreleased section of `CHANGES.md`

edlins commented:
Please hold, one or two small details to work out..

edlins commented:
Okay, I think this is good now.  To see how the rubber meets the road, take a look first at `PrecisionNumTest`.  Also, try running the `CompareNumTypes` example to see differences in time and accuracy as impacted by different precisions.

team172011 commented:
@edlins do you want to replace BigDecimal with PrecisionNum within this PR?

edlins commented:
> do you want to replace BigDecimal with PrecisionNum within this PR?

I'd prefer not to because this is already pretty big, and the replacement is also bigger than I thought and touches a lot of files in ways I was not expecting.  As soon as this is merged, I am happy to perform the replacement.

edlins commented:
I can work on the new Conflicts today or tomorrow.

team172011 commented:
> I'd prefer not to because this is already pretty big, and the replacement is also bigger than I thought and touches a lot of files in ways I was not expecting. As soon as this is merged, I am happy to perform the replacement.

okay

edlins commented:
That rebase was a little messy.  Plz double-check my manual merging.

edlins commented:
Fixed! (I think)

edlins commented:
Just wondering if anyone has had a chance to review this, please.

team172011 commented:
> Just wondering if anyone has had a chance to review this, please.
I will do, asap

team172011 commented on ta4j-core/src/main/java/org/ta4j/core/BaseBar.java:
```diff
@@ -230,7 +232,11 @@ public void addTrade(Num tradeVolume, Num tradePrice) {
+    public void addPrice(Num inPrice) {
+        Num price = inPrice;
+        if (numFunction != null) {
+            price = numFunction.apply(inPrice.getDelegate());
```
The new instance variable is only needed for this statement? If `numFunction != null` and the user tries to add another `Num ` implementation the `inPrice` will be converted to the `Num ` implementation of the `BaseBar` class. Is this the correct behaviour? The user can get unexpected rounding deviatations. I think we should not implicit transform different Num implementations, but throw a `ClassCastException ` if the user tries to mix up different `Num ` implementations. There are also default functions in `Bar` interface that support adding and transforming a `Number` or `String` into the corresponding `Num` implementation of the `BaseBar` class:
```
    default void addPrice(String price, Function<Number, Num> numFunction){
        addPrice(numFunction.apply(new BigDecimal(price)));
    }

    default void addPrice(Number price, Function<Number, Num> numFunction){
        addPrice(numFunction.apply(price));
}
```

edlins replied:
If `numFunction` is not stored and used then what is the purpose of the `numFunction` parameter in the constructor?
```
public BaseBar(Duration timePeriod, ZonedDateTime endTime, Function<Number, Num> numFunction)
```

team172011 replied:
You need the `numFunction` parameter to initialize some values:

```
this.volume = numFunction.apply(0);
this.amount = numFunction.apply(0);
```

edlins replied:
When I remove the change to `BaseBar` the following unit test in `PrecisionNumTest` (line 186) fails
```
assertEquals(32, ((BigDecimal) bigDecimalSeries.getBar(0).getClosePrice().getDelegate()).precision());
```
with
```
java.lang.AssertionError: expected:<32> but was:<120>
```

team172011 commented on ta4j-core/src/test/java/org/ta4j/core/TestUtils.java:
```diff
@@ -157,4 +160,38 @@ public static void assertNumNotEquals(double expected, Num actual) {
+                    actString = actString.substring(0, minLen) + "..";
+                throw new AssertionError(String.format("Failed at index %s: expected %s but actual was %s",
+                        i, expString, actString));
+            } else {
```
I think if the result is correct we do not have to log a message each time?

edlins commented:
> Looks good, I only have some minor requests

Thanks!  Responses added.

edlins commented:
LMK if I should remove or hack the failing unit test.

team172011 commented:
> LMK if I should remove or hack the failing unit test.
Why does it fail? 

Failed tests:   precisionNumTest(org.ta4j.core.num.PrecisionNumTest): expected:<32> but was:<120>
Tests run: 933, Failures: 1, Errors: 0, Skipped: 0

edlins commented:
> Why does it fail?

Because `BaseBar.AddPrice(Num)` doesn't use a `numFunction`.  It just overwrites the `Num` fields with the `price` parameter.  It should use a `numFunction`. This could probably be done implicitly with
```
closePrice = closePrice.numOf(price)
```
This would need to be done on all four assignments (open, close, max, min).

edlins commented:
> I think managing the Num logic in TimeSeries is the best way

Okay, I'll look at changing my test class to use `TimeSeries` instead of `Bar`. Do you want me to deprecate `Bar#addPrice` or just leave that alone for now?

edlins commented:
Found a `BaseBar` constructor that actually uses `numFunction.apply()` on `String` parameters.  It's hardcoded to a one-day `Duration` but that's how I wrote the tests originally so I lucked out.

team172011 commented:
> Do you want me to deprecate Bar#addPrice or just leave that alone for now?
We could also delete the addPrice function from Bar interface, but that would be much work in all exiting test cases. I have to think about it

edlins commented:
> We could also delete the addPrice function from Bar interface, but that would be much work in all exiting test cases. I have to think about it

Let's leave that out of scope for this PR.  I just added more test cases and a low precision (0.001) test.  The low precision was much less precise than `DoubleNum` and ran in 300% of the time.  I like `DoubleNum`.  I also added `assertIndiatorNotEquals(Indicator, Indicator, Num)` for negative testing.  I should probably add some test cases to `TestUtilsTest` but the tests in `PrecisionNumTest` cover it pretty well for now.

edlins commented:
FYI - this also implements #223 so that can be closed.
